cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/env/ContextStack.java

```diff
@@ -45,11 +46,17 @@ public Context peekContext() {
         return stack.peek();
     }
 
-    public void updateVariables(Map<String, Value> newVariables) {
-        Context flowContext = this.stack.getLast();
-        newVariables.keySet().stream()
-                .filter(variable -> flowContext.getVariable(variable) != null)
-                .forEach(variable -> flowContext.putVariable(variable, newVariables.get(variable)));
-
+    public Boolean updateVariables(Map<String, Value> newVariables) {
+        Context flowContext = peekContext();
+        AtomicReference<Boolean> changedContext = new AtomicReference<>(false);
+        newVariables.keySet()
+                .forEach(variable -> {
+                    if (flowContext.getVariable(variable) != null &&
+                            !flowContext.getVariable(variable).equals(newVariables.get(variable))) {
+                        flowContext.putVariable(variable, newVariables.get(variable));
+                        changedContext.set(true);
```

Comment:
There was no need to merge the filter clause and the forEach one, since it doesn't bring any performance improvement. It's just harder to read now.

Reply:
Move the conditions to a filter clause